xbmc/pvr/windows/GUIWindowPVRChannels.cpp

```diff
@@ -57,6 +57,16 @@ CGUIWindowPVRChannelsBase::~CGUIWindowPVRChannelsBase()
       this);
 }
 
+std::string CGUIWindowPVRChannelsBase::GetRootPath() const
+{
+  //! @todo make CGUIMediaWindow::GetRootPath() non-const.
```

Comment:
can you explain why?

Reply:
Sure. we are facing the old problem that somebody invents an interface - a virtual function - and makes an assumption about the implementation by defining the virtual function const. Now, the implementation has has either to provide the data from a member (eager evaluation enforced) or to do lazy evaluation, calculating the data on demand, but to be able to store the data in a member from within a const function, the member needs to be defined `mutable`. For this case here, I even need to call functions here doing the lazy data evaluation, which cannot be const by design. Thus, I have no other choice than to const_cast this.

There can be different opinions on this. One could say `GetRootPath` is **logically** const, thus the const_cast in the implementation is perfectly fine. I tend to say, the declaration of the virtual function needs to be revised for the reasons I gave above. There is no right and wrong, though. Mayby I should remove the comment.

Reply:
Thanks for explaining, I now understand the reason. Keeping the todo is fine to me, I'd rephrase it as the way it's written makes it sound like mandatory change and, as you say, there's not a right and wrong here (needs a careful evaluation).

Reply:
Okay, I will rephrase the comment to be more a question.

Reply:
Yeah I guess there's no right and wrong, but I tend to be in the pro const side here.
For me the name `GetRootPath` makes it obvious it should be const precisely to not have side effects like https://github.com/xbmc/xbmc/blob/4e291ab51b6b646fcbc110890592271d60d8120f/xbmc/pvr/windows/GUIWindowPVRBase.cpp#L465 being called.

At this time, `GetRootPath` modifies `m_vecItems` path, which imho is a brutal side effect for a "getter".